content/posts/2018-03-18-dependency-upgrade.md

```diff
@@ -0,0 +1,145 @@
+Actually, you will answer this not only to yourself but to your management, since such
+migrations usually consume a considerable amount of time which can be spent on evolving
+a product from the consumer perspective.
+
+Let’s split this vague question to simpler ones.
+
+* What benefits the migration will bring to the table?
+* How much time will it take?
+* How will it affect the development process?
+
+Answers like _the new one is just better_ do not work in real life.
+
+Unfortunately, the RxJava upgrade has a huge impact on a project,
+especially if it is practically based on RxJava and every component uses it this way or another.
+
+* There are some performance benefits, but the previous version is fine.
+* It will most likely take a huge amount of time (don’t forget to double it taking tests into account).
+* Developers have to adapt to new rules (no `null` usage at all is the biggest one).
+* Errors you make will most likely rise in runtime.
+
+It is always easier to sell huge performance boosts or improved development experience.
+Know all pros and cons. The truth might be harsh. Every developer wants to have nice new things,
+but when your backlog is filled with product-oriented tasks the reality kicks in.
+
+# Know Your Enemy
+
+* Take a closer look at changes and get really familiar with them.
+  Prepare to become a person who will be addressed with all rising questions and issues,
+  at least in the beginning.
+* Open a text editor of choice and start working on your RFC.
+  Put all pros and cons there, as well as a high-level changes overview.
+  You will be surprised, but not everyone on your team knows all details as well as you do.
+* Think about scheduling a meeting or something similar where you can set up
+  an improvised QA session. You will be surprised again, but your RFC might be
+  not as clear as you’ve imagined it in your mind.
+
+At this point, you might ask yourself an interesting question.
+
+> Why bother with all this team communication? I can upgrade everything myself!
+
+The answer is... teamwork!
+
+* Your changes will affect the codebase and the development process.
+* Everyone on the team will review your changes (hopefully), so it is always better to be on the same page.
+* As a bonus, there might be interesting thoughts or points you haven’t considered yourself.
+
+For example, RxJava brings [a huge amount of changes](https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0).
+The most notorious one is throwing the `NullPointerException` if a stream has `null` value in it.
+This amount of changes will affect everyone on your team, especially if you use RxJava heavily.
+
+Retrofit might be simpler in that regard even if [changes are not so small](https://github.com/square/retrofit/blob/master/CHANGELOG.md#version-200-2016-03-11).
+The reason is simple — scoping. Retrofit affects your network layer, but not every developer
```

Comment:
"reason is simple — scope of the dependency"?

Scoping sounds like a feature of library, like scoping in Dagger